Tratativas Envio/Recebimento, JSON, otimizações, correções e Dictionary de dados do Dataset

wellington1993 opened the pull request:
- Implementações relacionadas a montagem e o envio de JSON
- Tratativas em relação ao retorno XML
- Tratamentos para BLOB
- Ajustes nas tradução de valores para o servidor
- Ajustes gerais(tratativas e desempenho) com base nas tabelas testadas.

hotsoft-desenv4 committed:
Ticket_ID: #63381 - Refatorado para entregar Json

hotsoft-desenv7 committed:
Correção nome variavel e adicionando opção de "//hash//id" na leitura do IdRemoto no retorno do POST

hotsoft-desenv4 committed:
Ticket_ID: #63732 -
- Refatorações para guardar o nome dos detalhes (nomedetalheplural)

hotsoft-desenv4 committed:
Merge remote-tracking branch 'origin/MergeHotSoft' into MergeHotSoft

# Conflicts:
#	lib/DIntegradorModuloWeb.pas

hotsoft-desenv4 committed:
Ticket_ID: #63732 -
- Tratativas para codificação e decodificação de http/utf8

hotsoft-desenv4 committed:
Ticket_ID: #63732 -
- Tratativas para nomesingular e xmlcontent

hotsoft-desenv7 committed:
correções hibrido

hotsoft-desenv4 committed:
Ticket_Id: #63732 - Adaptar para buscar o datatype de forma otimizada

hotsoft-desenv7 committed:
ticket_id: #63381 - Correcao Hibrido JSON

hotsoft-desenv7 committed:
Hibrido - Alteração do TranslateValueToServer, alterando dataset para dictonary

hotsoft-desenv7 committed:
correção validação do field null para considerar campos vazios = ''

hotsoft-desenv4 committed:
Ajustes na nova implementação - DataSetToArray

Esta implementação foi necessária pois o Field passado por parâmetro no método TranslateValueToServer estava dando Access Violation ao acessar o Field.asString.

hotsoft-desenv7 committed:
Fazer o commit na base local utilizando dataset com parametros ao invés de um camomando SQL com "ExecSQL"

hotsoft-desenv2 committed:
Ticket_id #62469: Ajuste na conversao de dados irregulares do Banco

Signed-off-by: Wellington Torrejais da Silva <desenv2@labplus.com.br>

hotsoft-desenv7 committed:
removendo variavel não utilizada

wellington1993 commented:
@acras Seguem os commits para sua revisão.

acras committed:
clear para permitir limpar as translations

hotsoft-desenv4 committed:
Tratamento da unit System.JSon para compatibilidade com Delphi Berlin

hotsoft-desenv4 committed:
Tratamento da unit System.JSon para compatibilidade com Delphi Berlin

hotsoft-desenv4 committed:
Ticket_Id: #63997 - refatoração do híbrido para comportar os datatypes

hotsoft-desenv3 committed:
Ticket_ID: #63997 - Alteração para que não seja adicionado where no insert

hotsoft-desenv4 committed:
Ticket_Id: #63997 - Tratamento de PK quando for insert

hotsoft-desenv4 committed:
Merge remote-tracking branch 'origin/MergeHotSoft' into MergeHotSoft

hotsoft-desenv2 committed:
Ticket_ID: #63997 - Tratamento de erros no Get da sincronizacao

Signed-off-by: Wellington Torrejais da Silva <desenv2@labplus.com.br>

hotsoft-desenv2 committed:
Ticket_ID: #63997 - Tratamento de erros no Get da sincronizacao

Signed-off-by: Wellington Torrejais da Silva <desenv2@labplus.com.br>

hotsoft-desenv4 committed:
Ticket_Id: #63997 - Melhorias nos logs

hotsoft-desenv4 committed:
Ticket_Id: #64272 - implementação do soft delete -revisão de logs

hotsoft-desenv2 committed:
Ticket_id: #64272 - Removendo condicao desnecessaria do where de retorno do Idremoto

Signed-off-by: Wellington Torrejais da Silva <desenv2@labplus.com.br>

hotsoft-desenv2 committed:
Merge branch 'MergeHotSoft' of github.com:acras/hibrido_client into MergeHotSoft

hotsoft-desenv2 committed:
Ticket_id: #64272 - Ajuste na sintaxe de comparacao dos tipos de campo

Signed-off-by: Wellington Torrejais da Silva <desenv2@labplus.com.br>

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,16 +8,42 @@ interface
+    constructor Create(const aTableName: string; aDm : IDataPrincipal);
+    destructor Destroy; override;
+  end;
+
```
Estas classes guardam o tipo de dados para cada campo, este dicionário de dados servirá como apoio para gerar os parâmetros no update/insert.

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,16 +8,42 @@ interface
   TParamsType = (ptParam, ptJSON);
 
+  TAnonymousMethod = reference to procedure(aDataSet: TDataSet);
+
```
Este método anônimo para percorrer os registros do Dataset.

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -43,17 +69,53 @@   TTabelaDependente = record
   end;
 
-  TTabelaDetalhe = class
+  TJSONArrayContainer = class
```
Guarda o container com os dados para a geração do Json

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -43,17 +69,53 @@   TTabelaDependente = record
+    FFieldList : TFieldDictionaryList;
+    FDm : IDataPrincipal;
+  protected
+    function GetNomeTabela: string; virtual;
```
Properties se tornaram métodos para que possam ser acessadas pelas classes descendentes.

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -206,7 +280,18 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
     numRegistros := 0;
-    xmlContent := getRemoteXmlContent(url, http);
+
+    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), FOREGROUND_BLUE OR 4 ); //);
```
métodos para facilitar o debug com cores no modo console, podem ser refatorados.

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -323,17 +404,111 @@ function TDataIntegradorModuloWeb.getInsertStatement(node: IXMLDomNode): String;
-begin
-  beforeUpdateRecord(id);
-  dmPrincipal.execSQL(aUpdateStatement);
+procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const id: integer);
```
Este método faz o Update/Insert parametrizado

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -447,19 +617,35 @@ function TDataIntegradorModuloWeb.getFieldUpdateList(node: IXMLDomNode): string;
   result := result + getFieldAdditionalUpdateList(node);
 end;
 
+function TDataIntegradorModuloWeb.getFieldInsertList(node: IXMLDomNode): string;
```
Monta os campos de Insert, baseado no xml

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+end;
+
+procedure TDataIntegradorModuloWeb.SelectDetails(aValorPK: integer; aTabelaDetalhe: TTabelaDetalhe);
+var
```
Monta o Json das tabelas detalhe

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1167,4 +1566,191 @@ constructor TTabelaDetalhe.create;
+end;
+
+procedure TFieldDictionaryList.getTableFields;
+var
```
Busca os datatypes para popular o dicionário

hotsoft-desenv4 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -710,7 +1096,7 @@ function TDataIntegradorModuloWeb.saveRecordToRemote(ds: TDataSet;
             log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d, Erro: %s.',[ClassName, e.ErrorCode, Self.GetErrorMessage(e.ErrorMessage)])
           else if e.ErrorCode = 500 then
-            log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d. Erro: Erro interno no servidor. ',[ClassName, e.ErrorCode])
+            log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d. Erro: Erro interno no servidor: %s. ',[ClassName, e.ErrorCode, e.ErrorMessage])
```
tratamento de exception melhorado com mensagem de erro

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+  end;
+end;
+
+function TDataIntegradorModuloWeb.DataSetToArray(aDs: TDataSet) : TDatasetDictionary;
```
A função como o nome diz tem como objetivo converter um Dataset para um array bidimensional.

O motivo e que após uma quantidade de ciclos de sincronização a conversao de dados a partir de TField passava a nao trabalhar corretamente gerando erros de conversao.

Uma das teorias se tratava de que a origem fosse a conversao nao linear dos fields ou seja de modo que havia uma lógica sendo executada entre as conversoes do resultado.

Para solucionar o problemas foi feita a conversao dos valores do Field de forma sequencial e alterado os métodos para buscar os dados a partir da chave do Field do Dicionário.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+  BlobStream: TStringStream;
+  FieldStream: TStream;
+begin
+  Result := TDatasetDictionary.Create;
```
Foi feito o uso do Dicionário pois supria a estrutura de chave/valor necessária para uso nos outros métodos.
Como o TDictionary contem metodos de acesso e busca seu uso facilitou a implementação das funcoes.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+    nome := aDs.Fields[i].FieldName;
+    OutputDebugString(PWideChar(nome));
+
+    try
```
Aqui foi realizado um tratamento para erros de conversao causados geralmente por como os componentes identificam os tipos de dados com base nos tipos dos field no Firebird.

O uso do Variant traz a conversao dos dados para responsabilidade do compilador e nao do componente de forma mais primitiva o que resolveu algum problemas de conversao.

Em casos onde o compilador nao consegue tratar a conversao foi assumido o uso do AsString que dentre os testes em grande quantidade de registros foi o suficiente para suprir a necessidade dos outros métodos.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+
+    if (aDs.Fields[i].IsNull) or (fieldValue = '') then
+      value := ''
+    else if aDs.Fields[i].DataType = ftBlob then
```
Foi realizado também um tratamento diferenciado em fields do tipo Blob para extrair o DataSting do field.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+    end
+    else
+      value := fieldValue;
+    Result.Add(nome, value)
```
Adiciona o field com a chave sendo o nome dele e o valor o resultado de um tratamento de conversao do compilador usando variant tendo como fallback uma conversao do componente para AsString.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+  end;
+end;
+
+function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet;
```
Função que monta uma estrutura Json com base em um DataSet, um array bi-dimensional de valores e um array de translates.

Sendo resumidamente a iteração do dos translates onde a partir de cada nome de atributo/coluna se busca o valor no dicionário de valores.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+          aDs.fieldByName(aTranslations.get(i).pdv), aNestedAttribute, aTranslations.get(i).fkName, fieldValue);
+      if not JsonObjectHasPair(nome, Result) then
+        if Self.encodeJsonValues then
+          Result.AddPair(nome,  EncodeString(Trim(valor)))
```
Condicional para enviar os dados em Base64 o que evitou muitos problemas com encoding e binários.

hotsoft-desenv2 replied:
Mesmo se os dados estivessem sendo enviados via parâmetros na URL este tratamento para base64 seria importante.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+
+function TDataIntegradorModuloWeb.getNewId: Integer;
+begin
+  Result := 0;
```
Tratativa permitindo validacoes em outras funcoes.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +768,185 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+  Result := 0;
+end;
+
+procedure TDataIntegradorModuloWeb.addMasterTableToJson(aDs: TDataSet; apStream: TStringStream);
```
Monta o objeto pai do JSON com os objetos filhos presentes em uma estrutura de collection tipada.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -600,30 +965,62 @@ function TDataIntegradorModuloWeb.post(ds: TDataSet; http: TidHTTP; url: string)
+ if paramsType = ptJSON then
   begin
-    pStream := TStringStream.Create;
+    pStream := TStringStream.Create('', TEncoding.UTF8);
```
Importante: Sem algumas tratativas de Encoding no Stream os dados de retorno e envios podem vir com a codificação errada e salvar dados de forma errada. 

De modo que se observa uma boa pratica o uso explicito do Encoding no Stream.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -600,30 +965,62 @@ function TDataIntegradorModuloWeb.post(ds: TDataSet; http: TidHTTP; url: string)
-      //TODO buscar este pStrem j� como JSON
-      //Params.SaveToStream(pStream, TEncoding.UTF8);
+      Self.addDetailsToJsonList(ds);
+      Self.addMasterTableToJson(ds, pStream);
```
Adiciona o Json com os filhos n oStram para uso no POST

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,16 +8,42 @@ interface
   IDataPrincipalUnit, idURI, System.Classes, Windows,
   ISincronizacaoNotifierUnit, Data.SqlExpr,
-  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc;
+  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, HTTPApp, StrUtils,
```
O uso do Collections supre a necessidade de estruturar tipadas nos métodos.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,16 +8,42 @@ interface
   ISincronizacaoNotifierUnit, Data.SqlExpr,
-  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc;
+  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, HTTPApp, StrUtils,
+  Soap.EncdDecd, Variants {$IFDEF VER250}, Data.DBXJSON, Data.DBXPlatform {$ENDIF} {$IFDEF VER300}, System.JSON {$ENDIF};
```
Foi necessário o uso do EncdDecd para conversoes   para base64 de dados nao ansi.
Variants foi para uso nas conversoes. 

Condicionais para suprir o uso nas funcoes que dependem do JSON.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1167,4 +1566,191 @@ constructor TTabelaDetalhe.create;
+  Self.getTableFields;
+end;
+
+destructor TFieldDictionaryList.Destroy;
```
Método destrutor elegante para garantir a correta limpeza dos dados do componente.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -666,30 +1063,19 @@ function TDataIntegradorModuloWeb.saveRecordToRemote(ds: TDataSet;
-        finally
-          CoUninitialize;
-        end;
+        Result := Self.getXMLContentAsXMLDom(xmlContent);
```
Extraído código para função.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -600,30 +965,62 @@ function TDataIntegradorModuloWeb.post(ds: TDataSet; http: TidHTTP; url: string)
   end;
+end;
 
+function TDataIntegradorModuloWeb.GetIdRemoto(aDoc: IXMLDomDocument2): integer;
```
Centralização da lógica de busca de retorno do Array.

O retaguarda/backend mudou algumas formas de retorno no Rails 5 o que demandou varias mudanças ate se chegar em um padrão de retorno do XML, para suprir possíveis mudanças futuras este método foi separado.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -600,30 +965,62 @@ function TDataIntegradorModuloWeb.post(ds: TDataSet; http: TidHTTP; url: string)
+      Result := strToInt(aDoc.selectSingleNode('//hash//id').text)
+    else
+      Result := StrToInt(aDoc.selectSingleNode('objects').selectSingleNode('object').selectSingleNode('id').text);
+  except
```
Novo tratamento de exceção para casos onde o ID nao for recebido.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -989,26 +1372,42 @@ procedure TDataIntegradorModuloWeb.DataModuleCreate(Sender: TObject);
 
 function TDataIntegradorModuloWeb.translateValueToServer(translation: TNameTranslation;
-  fieldName: string; field: TField; nestedAttribute: string = ''; fkName: string = ''): string;
+  fieldName: string; field: TField; nestedAttribute: string = ''; fkName: string = ''; fieldValue: String = ''): string;
```
Agora a função trabalha tanto com o valor do Field e também com o valor já convertido vindo do Array.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -989,26 +1372,42 @@ procedure TDataIntegradorModuloWeb.DataModuleCreate(Sender: TObject);
+  if fieldValue <> '' then
+    ValorCampo := fieldValue
+  else
+    ValorCampo := field.AsString;
```
Se o valor nao vier convertido do Array usa o do Field

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -189,12 +262,13 @@ function TDataIntegradorModuloWeb.getObjectsList: string;
 procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
 var
-  url, xmlContent: string;
+  url, xmlContent, erro: string;
```
Tratamento adicional de erros pois nao havia retorno o que prejudicava a identificação dos problemas.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -206,7 +280,18 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
+    SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), 7);
+    xmlContent := getRemoteXmlContent(url, http, erro);
+
+    if (erro <> EmptyStr) then
```
Nao havia tratamento de retorno para erros de Get o que prejudicava a identificao dos problemas.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -206,7 +280,18 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
+    begin
+      vLog := Format('Erro importando "%s": "%s". '+ #13#10, [getHumanReadableName, GetErrorMessage(erro)]);
+      Self.Log(vLog);
+      raise EIntegradorException.Create(vLog);
```
Nao estava jogando a exceção para cima o que além de nao passar o erro impedia a captura dela por metidos acima deste.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -256,18 +342,13 @@ procedure TDataIntegradorModuloWeb.importRecord(node : IXMLDomNode);
-          insertRecord(statement);
+        try
+          qry := dmPrincipal.getQuery;
+          Self.updateInsertRecord(node, id);
```
Centralização da lógica em um método e uso do getQuery do dmPrincipal.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -507,13 +693,13 @@ function TDataIntegradorModuloWeb.translateFieldValue(
 begin
   result := QuotedStr(fieldValue);
-  if fieldType = 'integer' then
+  if (fieldType = 'integer') or (fieldType = 'float') then
```
Tratamentos adicionais de tipo de Field

hotsoft-desenv7 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -693,12 +707,19 @@ function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet; aTranslations: TT
     begin
       nome := aTranslations.get(i).server;
-      valor := translateValueToServer(aTranslations.get(i), aTranslations.get(i).pdv,
+      outputdebugstring(PwideChar(nome));
```
Foi usado apenas para depuração, pode ser removido mais tarde

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -785,25 +881,18 @@ procedure TDataIntegradorModuloWeb.addDetails(ds: TDataSet; params: TStringList)
-  finally
-    FreeAndNil(qry);
-  end;
+  RunDataSet(valorPk,
```
Iteração das tabelas detalhe a partir de uma procedure.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -582,6 +591,93 @@ procedure TDataIntegradorModuloWeb.OnWorkHandler(ASender: TObject; AWorkMode: TW
+    qry.Open;
+    while not qry.Eof do
+    begin
+      aProc(qry);
```
Executa a procedure que for passada via parâmetro.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -18,6 +18,8 @@   EIntegradorException = class(Exception)
   THttpAction = (haGet, haPost);
   TParamsType = (ptParam, ptJSON);
 
+  TAnonymousMethod = reference to procedure(aDataSet: TDataSet);
```
Declaração de procedure anônima.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -989,6 +1078,8 @@ procedure TDataIntegradorModuloWeb.DataModuleCreate(Sender: TObject);
   useMultipartParams := false;
   paramsType := ptParam;
   FstopOnPostRecordError := true;
+  Self.FDetailList := TDictionary<String, TJsonArray>.Create;
```
Lista tipada das tabelas detalhe.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -778,7 +777,9 @@ function TDataIntegradorModuloWeb.saveRecordToRemote(ds: TDataSet;
             if doc.selectSingleNode('//' + dasherize(nomeSingularSave) + '//id') <> nil then
               idRemoto := strToInt(doc.selectSingleNode('//' + dasherize(nomeSingularSave) + '//id').text)
-            else
+            else if doc.selectSingleNode('//hash//id') <> nil then
```
O retorno do Rails esta como Hash-> Id e nao como *nome*-Id
Entao foi feito este else

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -45,6 +45,14 @@   TTabelaDependente = record
 
+  TJSONArrayContainer = class
+  private
+    FJSonArray: TJsonArray;
```
Array de Items JSON

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -670,7 +670,7 @@ function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet; aTranslations: TT
           aDs.fieldByName(aTranslations.get(i).pdv), aNestedAttribute, aTranslations.get(i).fkName);
       if Self.encodeJsonValues then
-        Result.AddPair(nome, HTTPEncode(valor))
+        Result.AddPair(nome, UTF8Encode(HTTPEncode(valor)))
```
HttpEncode passa a ter também UTF8 Encode

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -883,7 +883,7 @@ function TDataIntegradorModuloWeb.GetErrorMessage(const aXML: string): string;
           node := list.FindNode('error');
           if node <> nil then
-            Result := node.Text;
+            Result := Utf8Decode(HTTPDecode(node.Text));
```
Da mesma forma que existe o encode de HTTP e de UTF-8 para envio tambem deve ser feito o decode do retorno de erro.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -734,12 +736,16 @@ function TDataIntegradorModuloWeb.GetIdRemoto(aDoc: IXMLDomDocument2): integer;
-    Result.loadXML(aXmlContent);
-  finally
-    CoUninitialize;
+  Result := nil;
```
Valor padrão de retorno inclusive para testa do retorno em outras funcoes.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1103,6 +1108,7 @@ procedure TDataIntegradorModuloWeb.DataModuleCreate(Sender: TObject);
   FstopOnPostRecordError := true;
   Self.FDetailList := TDictionary<String, TJSONArrayContainer>.Create;
   Self.encodeJsonValues := False;
+  translations.add('id', 'idremoto');
```
Este translate existia em todas as tabelas, entrao foi centralizado.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,7 +8,8 @@ interface
   ISincronizacaoNotifierUnit, Data.SqlExpr,
-  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, Data.DBXJSON, HTTPApp;
+  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, Data.DBXJSON, HTTPApp,
+  Soap.EncdDecd;
```
Encode de Base64 para suprir problemas de encoding e afins

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -672,7 +677,7 @@ function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet; aTranslations: TT
           aDs.fieldByName(aTranslations.get(i).pdv), aNestedAttribute, aTranslations.get(i).fkName);
       if Self.encodeJsonValues then
-        Result.AddPair(nome, UTF8Encode(HTTPEncode(valor)))
+        Result.AddPair(nome,  EncodeString(UTF8Encode(valor)))
```
Estava invertida a lógica do encoding o encode para UTF-8 deve ser em cima da String e nao do Base64

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -726,12 +731,19 @@ function TDataIntegradorModuloWeb.post(ds: TDataSet; http: TidHTTP; url: string)
+  except
+    on e: Exception do
+    begin
+      Self.log('Erro ao ler Campo "ID" no XML de retorno, Tabela: ' + nomeTabela + ' - ' + e.Message, 'Sync');
```
Tratamento de erro adicional para situacoes onde nao existe retorno do ID.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1130,6 +1142,7 @@ function TDataIntegradorModuloWeb.translateValueToServer(translation: TNameTrans
 var
   lookupIdRemoto: integer;
   fk: string;
+  StringStream: TStringStream;
```
Tratamento adicional para dados Binários

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1163,8 +1190,10 @@ function TDataIntegradorModuloWeb.translateValueToServer(translation: TNameTrans
   fk: string;
   StringStream: TStringStream;
 begin
+  result := field.asString;
```
Valor padrão para Result, por definição trata como String.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1173,14 +1202,10 @@ function TDataIntegradorModuloWeb.translateValueToServer(translation: TNameTrans
       lookupIdRemoto := dmPrincipal.getSQLIntegerResult('SELECT idRemoto FROM ' +
         translation.lookupRemoteTable +
-        ' WHERE ' + fk + ' = ' + field.AsString);
+        ' WHERE ' + fk + ' = cast(' + field.AsString + ' as blob)' );
```
Para suprir problemas com buscas tornando a mesma mais abrangente dado o uso do cast para binário.
Foi realizado testes e em alguns registros auxiliou na busca.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -460,27 +460,21 @@ procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const i
-                  FreeAndNil(BlobStream);
-                end;
-              end
+              case Field.DataType of
```
Melhoria estrutural do código pois estava dificultando a depuração e a leitura.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1259,7 +1267,8 @@ procedure TDataIntegradorModuloWeb.postRecordsToRemote(http: TidHTTP = nil);
       if notifier <> nil then
         notifier.unflagSalvandoDadosServidor;
-      Self.log('Commitando post de records para remote. Classe: ' + ClassName, 'Sync')
+      if Total > 0 then
```
Somente logar que esta comitando se houver o que ser comitado.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -487,7 +487,19 @@ procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const i
+    try
+      qry.ExecSQL;
+    except
+      on E:Exception do
```
Tratamento de exceção pois nao havia retorno do que causava o problema dificultando a correção.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -487,7 +487,19 @@ procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const i
+        ParamLog := EmptyStr;
+        for i := 0 to qry.Params.Count - 1 do
+          ParamLog := ParamLog + qry.Params[i].Name + ' = "' + qry.Params[i].AsString + '"' + #13#10;
+        vLog := 'Erro ExecSQL: ' + #13#10 + qry.CommandText + #13#10 + ParamLog + #13#10 + e.Message;
```
Tratamento detalhado possibilitando a correção do problema.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -804,13 +816,21 @@ procedure TDataIntegradorModuloWeb.RunDataSet(const aValorPK: integer; aTabelaDe
+        aProc(qry);
+        qry.Next;
+      end;
+    except
```
Tratamento de exceção de modo que erros nao passem sem detalhamento e tratamento.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -281,8 +281,15 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
+      if (erro <> EmptyStr) then
+      begin
+        log := Format('Erro importando "%s": "%s". '+ #13#10, [getHumanReadableName, GetErrorMessage(erro)]);
+        Self.DataLog.log(log);
```
Tratamento adicional de erro

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -281,8 +281,15 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
+      begin
+        log := Format('Erro importando "%s": "%s". '+ #13#10, [getHumanReadableName, GetErrorMessage(erro)]);
+        Self.DataLog.log(log);
+        raise EIntegradorException.Create(log);
```
Jogando o erro para cima para que o mesmo nao seja desconsiderado.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1110,7 +1113,7 @@ function TDataIntegradorModuloWeb.GetErrorMessage(const aXML: string): string;
           node := list.FindNode('error');
           if node <> nil then
-            Result := Utf8Decode(HTTPDecode(node.Text));
+            Result := UTF8ToString(HTTPDecode(node.Text));
```
Ajuste de Deprecation na funcao de Decode de UTF-8

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -279,7 +279,10 @@ procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
-    xmlContent := getRemoteXmlContent(url, http);
+
+    xmlContent := getRemoteXmlContent(url, http, erro);
+    if Self.DataLog <> nil then
```
Evita chamada se DataLog nao estiver definido.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -262,7 +262,7 @@ function TDataIntegradorModuloWeb.getObjectsList: string;
 procedure TDataIntegradorModuloWeb.getDadosAtualizados(http: TIdHTTP = nil);
 var
-  url, xmlContent, erro: string;
+  url, xmlContent, erro, log: string;
```
Log adicional para tratar e nao ignorar erros no GET

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -207,7 +207,7 @@   TDataIntegradorModuloWeb = class(TDataModule)
     function getUpdateStatement(node: IXMLDomNode; const id: integer): String; virtual;
     function getInsertStatement(node: IXMLDomNode): String; virtual;
-    function getNewId: Integer; virtual; abstract;
+    function getNewId: Integer; virtual;
```
Para que se possa testar resultado/implemetacao da função e se houve retorno/problema.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -405,6 +405,7 @@ procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const i
   BlobStream: TStringStream;
   Existe: Boolean;
   Field: TFieldDictionary;
+  NewId: integer;
```
Sem ID nao há como inserir o registro pois o Firebird nao contempla auto increment

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -421,7 +422,12 @@ procedure TDataIntegradorModuloWeb.updateInsertRecord(node: IXMLDomNode; const i
     else
     begin
       FieldsListInsert := self.getFieldInsertList(node);
+      NewId := Self.getNewId;
```
Chamada a função para buscar o ID para inserção.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -8,8 +8,8 @@ interface
-  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, Data.DBXJSON, HTTPApp,
-  Soap.EncdDecd, Variants;
+  Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, HTTPApp,
+  Soap.EncdDecd, Variants {$IFDEF VER250}, Data.DBXJSON {$ENDIF} {$IFDEF VER300} System.JSON {$ENDIF};
```
Retrocompatibilidade para versoes diferente do Delphi,

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1320,6 +1345,8 @@ destructor TDataIntegradorModuloWeb.Destroy;
   for Item in Self.FDetailList do
     Item.Value.Free;
   Self.FDetailList.Free;
+  if Self.FFieldList <> nil then
```
Garante a limpeza dos item após o uso.

hotsoft-desenv2 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -9,7 +9,7 @@ interface
   ISincronizacaoNotifierUnit, Data.SqlExpr,
   Xml.XMLIntf, Winapi.ActiveX, XML.XMLDoc, System.Generics.Collections, Data.DBXJSON, HTTPApp,
-  Soap.EncdDecd;
+  Soap.EncdDecd, Variants;
```
Para suprir o uso do Variant nos tratamentos de conversao

hotsoft-desenv4 committed:
Ticket_Id: #64272 -

- Tratativa ao importar um registro, para exibir o log mais verboso
- Tratativa no método updateInsertRecord para não inserir o campo PK caso já tenha sido inserido. Também tratado para os campos Data e Currency

hotsoft-desenv4 committed:
Ticket_Id: #64305 -

- Tratativa no método getUpdateData para startar a transação a cada tabela (e não a cada operação de get)
- Também tratado para exibir o log de erros.

hotsoft-desenv4 committed:
Merge remote-tracking branch 'origin/MergeHotSoft' into MergeHotSoft

# Conflicts:
#	lib/DIntegradorModuloWeb.pas

hotsoft-desenv4 committed:
Ticket_Id: #64305 -

- Soft delete - implementação das classes que guardam o dicionário com as traduções da tabela local versus tabela no servidor.

hotsoft-desenv4 committed:
Ticket_Id: #64305 -

- Soft delete - Implementação para fazer o softdelete nos gets

hotsoft-desenv4 committed:
Ticket_Id: #64305 - Transações Commit e RollBack retornadas

hotsoft-desenv4 committed:
Ticket_Id: #64305 - Correção de memory leak

hotsoft-desenv4 committed:
Ticket_Id: #64305 -
- GetNewID recebe o XMLNode como referencia
- Isto foi feito para implementar o GET das classes cuja PK também é PK da tabela Empresa (LocalAtendimento, Fornecedor, Laboratorio, Operadora, Apoiado, Estabelecimento)

hotsoft-desenv4 committed:
Ticket_Id: #64272 -  retirada a regra que apaga o :deleted (regra específica HS)

hotsoft-desenv4 committed:
Ticket_Id: #64272 -  assinatura do método BeforeUpdateInsert alterada com o parâmetro Handled.

hotsoft-desenv4 committed:
Ticket_Id: #64594 - Get mestre e detalhe

hotsoft-desenv4 committed:
Ticket_id: #64595 -

- Refatorações:
   - TDatasetDictionary e TTranslateTableNames se tornaram TStringDictionary
   - método Destroy da classe TJsonArrayContainer para resolver memory leaks
   - criado a classe TDetailList para resolver memory leaks
   - property StopOnGetRecordError: caso seja True, para o get da tabela atual e vai para a próxima tabela do array de Getters
   - ImportRecord se tornou function para comportar a alteração da nova property StopOnGetRecordError
   - métodos translateFieldValue e translateValueFromServer recebem o integrador como parâmetro
   - método BeforeUpdateRecord retirado
   - Criado o método SetQueryParameters para organização do código
   - refatorado para utilizar o construtor (Create) da Classe TDataIntegradorModuloWeb e não mais o método DataModuleCreate
   - método getDadosAtualizados implementado para comportar a nova property StopOnGetRecordError
   - Revisados os pontos de criação de StringStream para comportar encoding UTF8
   - Revisados os pontos de commit que ficavam com transação aberta, sem rollback
   - objeto stream no método saveRecordToRemote estava com memory leak

- TXMLNodeDictionary: dicionário para guardar as tabelas filhas e seus respectivos nós

hotsoft-desenv7 committed:
correção do EncodeBase64, em alguns campos o encode não estava sendo feito de uma forma correta, alterado para usar o Encode do componente Soap

hotsoft-desenv4 committed:
Ticket_Id: #64594 - Ajustes na variável para UTF8String (memory leak) e tratamento no data type dos campos

hotsoft-desenv4 committed:
Merge remote-tracking branch 'origin/MergeHotSoft' into MergeHotSoft

hotsoft-desenv7 committed:
Realizar POST e GET de arquivos binários salvos em campos BLOB Binary

hotsoft-desenv7 committed:
Merge remote-tracking branch 'origin/MergeHotSoft' into MergeHotSoft

# Conflicts:
#	lib/DIntegradorModuloWeb.pas

hotsoft-desenv7 committed:
Descomentando o método POST

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -434,17 +434,16 @@ function TDataIntegradorModuloWeb.getTranslatedTable(const aServerName: string):
 var
   i: integer;
-  ValorCampo: string;
+  ValorCampo: UTF8String;
```
Para resolver problemas como access violation na Query.
http://docwiki.embarcadero.com/Libraries/Seattle/en/System.UTF8String

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -998,13 +1003,17 @@ function TDataIntegradorModuloWeb.DataSetToArray(aDs: TDataSet) : TStringDiction
+        try
+          Input.LoadFromStream(FieldStream);
+          Input.Position := 0;
+          Value := TIdEncoderMIME.EncodeStream(Input);
```
Foi trocada a biblioteca utilizada no Encoding para Base64

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1037,8 +1046,12 @@ function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet;
-          Result.AddPair(nome, EncodeBase64(PAnsiChar(StringUTF8), Length(StringUTF8)))
+        begin
+          if aDs.fieldByName(aTranslations.get(i).pdv).DataType = ftblob then
+            Result.AddPair(nome, valor)
```
Já veio codificado neste ponto

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -1037,8 +1046,12 @@ function TDataIntegradorModuloWeb.getJsonObject(aDs: TDataSet;
+          if aDs.fieldByName(aTranslations.get(i).pdv).DataType = ftblob then
+            Result.AddPair(nome, valor)
+          else
+            Result.AddPair(nome, TIdEncoderMIME.EncodeString(StringUTF8, IndyTextEncoding_UTF8))
```
Se não for binário, codifica para base64 garantindo encoding

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -255,7 +256,7 @@   TTabelaDetalhe = class(TDataIntegradorModuloWeb)
 implementation
 
-uses AguardeFormUn, ComObj;
+uses AguardeFormUn, ComObj, idCoderMIME, IdGlobal;
```
Usado o Indy para fazer encoding para Base64

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -43,30 +72,33 @@   TTabelaDependente = record
+
+  TDetailList = class(TDictionary<String, TJSONArrayContainer>)
+  public
+    destructor Destroy; override;
```
Garante a correta destruição dos filhos evitando memory leaks

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -43,30 +72,33 @@   TTabelaDependente = record
-    constructor create;
+    nomePkLocal: string;
+    function getJsonArray: TJsonArray;
+    destructor Destroy; override;
```
Evita memory leaks destruindo os filhos do objeto

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -43,30 +72,33 @@   TTabelaDependente = record
-    procedure SetdmPrincipal(const Value: IDataPrincipal);
-    function getdmPrincipal: IDataPrincipal;
-
+    FStopOnGetRecordError : boolean;
```
Flag que especifica se a sincronização deve ou não continuar com a inserção ao obter erro em um dos inserts do processo de GET

wellington1993 commented on lib/DSincronizadorModuloWeb.pas:
```diff
@@ -348,6 +368,7 @@ procedure TRunnerThreadPuters.Execute;
         dm := nil;
         if http <> nil then
           FreeAndNil(http);
+        FreeAndNil(lTranslateTableNames);
```
Ajuste de Memory Leak

wellington1993 commented on lib/DSincronizadorModuloWeb.pas:
```diff
@@ -163,12 +165,21 @@ procedure TDataSincronizadorModuloWeb.getUpdatedData;
     finally
       dm := nil;
-      http := nil;
+      FreeAndNil(http);
```
Ajuste de memory leak

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -764,7 +1317,7 @@ function TDataIntegradorModuloWeb.GetErrorMessage(const aXML: string): string;
           node := list.FindNode('error');
           if node <> nil then
-            Result := node.Text;
+            Result := UTF8ToString(HTTPDecode(node.Text));
```
Garante que o resultado venha decodificado e que o encoding seja ajustado.

wellington1993 commented on lib/DIntegradorModuloWeb.pas:
```diff
@@ -710,7 +1263,7 @@ function TDataIntegradorModuloWeb.saveRecordToRemote(ds: TDataSet;
             log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d, Erro: %s.',[ClassName, e.ErrorCode, Self.GetErrorMessage(e.ErrorMessage)])
           else if e.ErrorCode = 500 then
-            log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d. Erro: Erro interno no servidor. ',[ClassName, e.ErrorCode])
+            log := Format('Erro ao tentar salvar registro. Classe: %s, C�digo de erro: %d. Erro: Erro interno no servidor: %s. ',[ClassName, e.ErrorCode, e.ErrorMessage])
```
Não adianta logar se não exibir a mensagem do erro de origem no servidor.

hotsoft-desenv7 committed:
ticket_id: #64594 movendo leitura de campos ftFMTBcd de string para float

hotsoft-desenv4 committed:
Ticket_Id: #64594 - correções no campo Float

hotsoft-desenv4 committed:
Ticket_Id: #64947 - correções de memory leak, alteração de array para TObjectList

hotsoft-desenv4 committed:
Ticket_Id: #64947 - Log no Get

hotsoft-desenv7 committed:
ticket_id: #64947 - correção tabelas detalhes

hotsoft-desenv7 committed:
ticket_id: #64947 - correção do underline no GET

hotsoft-desenv4 committed:
Ticket_Id: #64947 - Ajustado para fazer o update do campo salvouretaguarda ao final de cada Get.

hotsoft-desenv4 committed:
Ticket_Id: #64947 -
- Refatoração da busca do WHERE (quando for duasVias)
- Ao final do get, atualizar o versionId com o último version id

hotsoft-desenv7 committed:
ticket_id: #64947 - update do salvouretaguarda apenas para tabela detalhe

hotsoft-desenv4 committed:
Ticket_Id: #64947 -
- Alterado para trabalhar com a função
   function GetDefaultValueForSalvouRetaguarda: Char; virtual;

hotsoft-desenv7 committed:
ticket_id: #64947 - correç~eos hibrido

hotsoft-desenv4 committed:
Ticket_Id: #64947 - ao verificar se o campo já existe, faltou concatenar com a vírgula

hotsoft-desenv4 committed:
Ticket_Id: #64947 -
- Popular as traduções das tabelas primeiro
- atualizar o maior version_id com o último version id recebido no get.

hotsoft-desenv4 committed:
Ticket_Id: #64947 -
- Retirado o count(*), foi subido por engano

hotsoft-desenv4 committed:
Ticket_Id: #64947 -
- resincronizar o registro caso dê algum problema no Get

hotsoft-desenv7 committed:
ticket_id: #64947 - não estava incluindo corretamente as tabelas detalhes a lista do translate

hotsoft-desenv7 committed:
ticket_id: #64947 - não estava incluindo corretamente as tabelas detalhes a lista do translate

hotsoft-desenv4 committed:
Ticket_Id: #64693
- híbrido - novo endereço de sincronização - só escrever logs quando for hibridodll

hotsoft-desenv7 committed:
correção tabela detalhe

hotsoft-desenv7 committed:
ticket_id: #64947 - removendo Version_id do list field do POST

hotsoft-desenv7 committed:
Somente atualizar registros que não tiveram alteração no intervalo da sincronização, dessa forma evita que um dado recem alterado seja substituido por um pouco mais antigo.
